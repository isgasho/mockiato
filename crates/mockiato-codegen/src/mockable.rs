use crate::generate::{generate_mock, GenerateMockOptions};
use crate::parse::mockable_attr::MockableAttr;
use crate::parse::trait_decl::TraitDecl;
use crate::spanned::SpannedUnstable;
use crate::Error;
use proc_macro::{Diagnostic, Level, Span, TokenStream};
use quote::quote;
use syn::{AttributeArgs, Item, ItemTrait};

#[derive(Default)]
pub(crate) struct Mockable;

impl Mockable {
    pub(crate) fn new() -> Self {
        Self::default()
    }

    pub(crate) fn expand(&self, attr: AttributeArgs, item: Item) -> TokenStream {
        let item_backup = item.clone();

        #[doc(hidden)]
        macro try_or_return($expr: expr) {
            match $expr {
                Ok(value) => value,
                Err(_) => return TokenStream::from(quote! { #item_backup }),
            }
        }

        let mockable_attr = try_or_return!(MockableAttr::parse(attr).map_err(Error::emit));
        let item_trait = try_or_return!(extract_item_trait(item).map_err(Error::emit));
        let trait_decl = try_or_return!(TraitDecl::parse(item_trait.clone())
            .map_err(|err| err
                .emit_with(|d| d.span_note(Span::call_site(), "Required for mockable traits"))));

        let generated_mock = generate_mock(
            &trait_decl,
            GenerateMockOptions {
                custom_struct_ident: mockable_attr.name_attr.map(|attr| attr.ident),
                force_static_lifetimes: mockable_attr.static_attr.is_some(),
            },
        );

        TokenStream::from(quote! {
            #item_trait
            #generated_mock
        })
    }
}

fn extract_item_trait(item: Item) -> Result<ItemTrait, Error> {
    match item {
        Item::Trait(item_trait) => Ok(item_trait),
        _ => Err(Error::Diagnostic(
            Diagnostic::spanned(
                item.span_unstable(),
                Level::Error,
                "Only traits can be made mockable",
            )
            .span_note(Span::call_site(), "Required because of this attribute"),
        )),
    }
}
