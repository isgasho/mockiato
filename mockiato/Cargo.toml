cargo-features = ["edition"]

[package]
name = "mockiato"
version = "0.1.0"
authors = ["Ruben Schmidmeister <ruben.schmidmeister@icloud.com>"]
edition = "2018"

[dependencies]
mockiato_codegen = { path = "../mockiato_codegen" }
