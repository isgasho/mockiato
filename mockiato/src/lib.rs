mod arguments;
mod call;
mod expected_calls;
mod matcher;
pub mod return_value;

pub use self::arguments::*;
pub use self::call::*;
pub use self::expected_calls::*;
pub use self::matcher::*;
