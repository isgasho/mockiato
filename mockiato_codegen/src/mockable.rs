use std::sync::RwLock;
use syntax::ast::{self, Ident, VariantData, DUMMY_NODE_ID};
use syntax::ext::base::{Annotatable, ExtCtxt, MultiItemDecorator};
use syntax::ext::build::AstBuilder;
use syntax::ptr::P;
use syntax_pos::Span;

use crate::parse::mockable_attr::MockableAttr;
use crate::parse::name_attr::NameAttr;
use crate::parse::trait_bounds::TraitBounds;
use crate::parse::trait_decl::TraitDecl;
use crate::definition_id::DefId;
use crate::trait_bound_resolver::{TraitBoundResolver, TraitBoundType};

pub(crate) struct Mockable {
    trait_bound_resolver: RwLock<Box<dyn TraitBoundResolver>>,
}

const TRAIT_BOUND_RESOLVER_ERR: &str = "Internal Error: Trait Bound Resolver is poisoned";

impl Mockable {
    pub(crate) fn new(trait_bound_resolver: Box<dyn TraitBoundResolver>) -> Self {
        Self {
            trait_bound_resolver: RwLock::new(trait_bound_resolver),
        }
    }

    fn register_current_trait(&self, trait_bound_def_id: DefId, trait_decl: &TraitDecl) {
        self.trait_bound_resolver
            .write()
            .expect(TRAIT_BOUND_RESOLVER_ERR)
            .register_mocked_trait(trait_bound_def_id, &trait_decl);
    }

    fn mock_trait_bound_impls(&self, cx: &mut ExtCtxt, trait_decl: &TraitDecl) {
        let trait_bounds = TraitBounds::parse(trait_decl).0;
        for trait_bound in trait_bounds {
            let identifier = trait_bound.identifier;
            let trait_bound_resolver = self
                .trait_bound_resolver
                .read()
                .expect(TRAIT_BOUND_RESOLVER_ERR);
            let trait_bound_type = trait_bound_resolver
                .resolve_trait_bound(&identifier);
            self.mock_trait_bound_type(cx, trait_bound.span, &trait_bound_type);
        }
    }

    fn mock_trait_bound_type(&self, cx: &mut ExtCtxt, sp: Span, trait_bound_type: &Option<TraitBoundType>) {
        match *trait_bound_type {
            None => {
                cx.parse_sess
                .span_diagnostic
                .mut_span_err(sp, "The referenced trait has has not been marked as #[mockable] or doesn't exist")
                .help("Mockable traits are handled from top to bottom, try declaring this trait earlier.")
                .emit();
            },
            _ => {}
        }
    }
}

impl MultiItemDecorator for Mockable {
    fn expand(
        &self,
        cx: &mut ExtCtxt,
        _sp: Span,
        meta_item: &ast::MetaItem,
        item: &Annotatable,
        push: &mut dyn FnMut(Annotatable),
    ) {
        let trait_decl = match TraitDecl::parse(cx, item) {
            Ok(trait_decl) => trait_decl,
            Err(_) => return,
        };

        let mockable_attr = match MockableAttr::parse(cx, meta_item) {
            Some(mockable_attr) => mockable_attr,
            None => return,
        };

        // self.register_current_trait(ruben_give_id, &trait_decl);
        
        let mock_struct_ident = mock_struct_ident(&trait_decl, mockable_attr.name_attr);

        let mut mock_struct = cx
            .item_struct(
                meta_item.span,
                mock_struct_ident,
                VariantData::Unit(DUMMY_NODE_ID),
            ).into_inner();

        if let Some(derive_attr) = mockable_attr.derive_attr {
            mock_struct.attrs.push(derive_attr.expand(cx));
        }

        push(Annotatable::Item(P(mock_struct)));

        // self.mock_trait_bound_impls(&mut cx, &trait_decl);
    }
}

fn mock_struct_ident(trait_decl: &TraitDecl, name_attr: Option<NameAttr>) -> Ident {
    name_attr
        .map(|attr| attr.expand())
        .unwrap_or_else(|| Ident::from_str(&format!("{}Mock", trait_decl.ident)))
}
