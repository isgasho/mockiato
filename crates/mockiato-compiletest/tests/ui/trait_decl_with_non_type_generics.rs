use mockiato::mockable;

#[mockable]
trait Foo<'a> {}

fn main() {}
