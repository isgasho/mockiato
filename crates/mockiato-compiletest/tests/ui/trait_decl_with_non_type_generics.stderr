error: Lifetimes are not supported on mockable traits
 --> $DIR/trait_decl_with_non_type_generics.rs:4:25
  |
4 | trait TraitWithLifetime<'a> {}
  |                         ^^
  |
note: Required for mockable traits
 --> $DIR/trait_decl_with_non_type_generics.rs:3:1
  |
3 | #[mockable]
  | ^^^^^^^^^^^

error: Lifetimes are not supported on mockable traits
 --> $DIR/trait_decl_with_non_type_generics.rs:7:40
  |
7 | trait TraitWithLifetimeAndGenericParam<'a, B> {}
  |                                        ^^
  |
note: Required for mockable traits
 --> $DIR/trait_decl_with_non_type_generics.rs:6:1
  |
6 | #[mockable]
  | ^^^^^^^^^^^

error: aborting due to 2 previous errors

